Add more sensors for some Tuya-based pool devices

IDisposable opened the pull request:
## Proposed change
The common Tuya-supported devices for Pool Chemistry Tester wnd the Smart Pool Chlorinator are exposed using the normal Tuya interfaces. Adding support for the sensors, switches and settings should be able to use the settings found in the API cloud desktop.


## Type of change

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [X] New feature (which adds functionality to an existing integration)
- [ ] Deprecation (breaking change to happen in the future)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Additional information
This is my first attempt at adding things to an integration, and Tuya is a bit unique in the way it works. I likely got it all wrong. I also have never built HomeAssistant (I just run on an HA Green), so I have no idea how to actually test this, looking for pointers.

## Checklist

- [ ] The code change is tested and works locally.
- [x] Local tests pass. **Your PR cannot be merged unless tests pass**
- [X] There is no commented out code in this PR.
- [X] I have followed the [development checklist][dev-checklist]
- [X] I have followed the [perfect PR recommendations][perfect-pr]
- [X] The code has been formatted using Ruff (`ruff format homeassistant tests`)
- [ ] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [ ] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [ ] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [ ] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [ ] For the updated dependencies - a link to the changelog, or at minimum a diff between library versions is added to the PR description.


To help with the load of incoming pull requests:

- [ ] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone+-status%3Afailure

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/development_checklist/
[manifest-docs]: https://developers.home-assistant.io/docs/creating_integration_manifest/
[quality-scale]: https://developers.home-assistant.io/docs/integration_quality_scale_index/
[docs-repository]: https://github.com/home-assistant/home-assistant.io
[perfect-pr]: https://developers.home-assistant.io/docs/review-process/#creating-the-perfect-pr

IDisposable committed:
Add explicit pH unit

Make ph (potential of hydrogen) a proper unit

IDisposable committed:
Add total dissolved solids (TDS)

IDisposable committed:
Add concentration sensor device class

IDisposable committed:
Add additional Tuya MAC prefixes

IDisposable committed:
Add concentration for Grams per Liter

IDisposable committed:
Add support for Pool Chlorinators

Most Tuya-based pool chlorinators act the same way, add support for all the switches. settings, and sensors.

IDisposable committed:
Add support for Pool Testers

Most Tuya-based pool testers act the same way, add support for all the settings, and sensors.

# Conflicts:
#	homeassistant/components/tuya/const.py
#	homeassistant/components/tuya/icons.json

IDisposable commented:
This is the capture of the attributes from the Tuya developer portal for my two devices.

[PoolTester.json](https://github.com/user-attachments/files/16721499/PoolTester.json)
[PoolChlorinator.json](https://github.com/user-attachments/files/16721500/PoolChlorinator.json)

IDisposable commented:
Would it be possible to approve the workflows so I can see if I broke anything? @tuya, @zlinoliver, @frenck

MartinHjelmare commented on homeassistant/components/sensor/const.py:
```diff
@@ -139,6 +140,12 @@ class SensorDeviceClass(StrEnum):
     Unit of measurement: `ppm` (parts per million)
     """
 
+    CONCENTRATION = "concentration"
```
This PR is changing our entity model. Before we can review this PR there needs to be approval in a discussion in our architecture repository.

https://github.com/home-assistant/architecture/discussions

https://developers.home-assistant.io/docs/core/entity#changing-the-entity-model

Also, please don't change the entity model in the same PR as changing an integration.

IDisposable replied:
I will create a follow up PR for the entity model changes after we discuss this integration. I have [committed and pushed a change](https://github.com/home-assistant/core/pull/124459/commits/40e9875ca84976800f07caaeaaf7f191fbe17918) that doesn't require the entity model changes and leaves pH as unitless, doesn't add CONCENTRATION, doesn't add TOTAL_DISSOLVED_SOLIDS and thus uses different types underneath for the new sensors/numbers.

IDisposable committed:
Revert schema changes

Splitting the addition of the integration for the new Tuya sensors from the changes in the data model.
Removes: POTENTIAL_OF_HYDROGEN and pH UNIT, reverts DEVICE_CLASS_UNIT[SensorDeviceClass.PH] to unitless.
Removes: CONCENTRATION SensorDeviceClass.
Removes SensorDeviceClass.TOTAL_DISSOLVED_SOLIDS
Replace: TOTAL_DISSOLVE_SOLIDS with simple CONCENTRATION
Also fixed mistype of ec_warning_min (should have been CONDUCTIVITY)
Also fixed type for SENSOR specific_gravity to use the correct SensorDeviceClass.CONCENTRATION.

IDisposable commented:
[Pushed a commit](https://github.com/home-assistant/core/pull/124459/commits/40e9875ca84976800f07caaeaaf7f191fbe17918) that backs out the entity model changes (we can always revisit those after the discussion). Ready for integration-only review.

IDisposable committed:
Remove unintentional blank line

IDisposable committed:
Remove rest of CONCENTRATION references

Also turns out min_value, max_value, and step are not supported in Tuya (unlike other places) so removed those.
Fixed formatting issue.

IDisposable commented:
Fixed the build errors by eliminating all the remaining unsupported stuff and fixed the format issue.

MartinHjelmare commented:
Please don't mark the PR ready for review until you've checked that linting and tests are ok in your local development environment.

IDisposable committed:
Fix RUFF issues.

# Conflicts:
#	homeassistant/components/tuya/const.py

IDisposable commented:
Any pointers on fixing the [`hassfest` build warning](https://github.com/home-assistant/core/actions/runs/10589884491/job/29344830723?pr=124459)?

IDisposable committed:
Add myself to code owners.

MartinHjelmare commented:
> Any pointers on fixing the [hassfest build warning](https://github.com/home-assistant/core/actions/runs/10589884491/job/29344830723?pr=124459)?

It says what you should do in the build error.

IDisposable commented:
It gives an error if I run that command...

IDisposable committed:
Run hassfest

IDisposable commented:
Turns out (by looking in the `.pre-commit-config.yaml` that the command should be `script/run-in-env.sh python3 -m script.hassfest` to ensure the environment is properly setup (from within the DevContainer)

MartinHjelmare commented:
The CI build passed. Good! Please wait for a review from a codeowner.

IDisposable commented:
I'm still waiting on a code-review, I would LOVE for this to be merged please.

MartinHjelmare commented:
To avoid any reasons for not getting a review, you can rebase your branch and solve the current merge conflicts.

IDisposable commented:
Rebased to dev

frenck commented on homeassistant/components/tuya/select.py:
```diff
@@ -307,6 +307,13 @@
             entity_category=EntityCategory.CONFIG,
         ),
     ),
+    # Early Product "QT" device (for Pool Chlorinator)
+    "qt": (
```
As far as I am aware, we shouldn't be adding "qt" devices. They can be anything, are considered legacy, or unsupported by Tuya.

frenck commented on homeassistant/const.py:
```diff
@@ -1349,6 +1349,7 @@ class UnitOfPrecipitationDepth(StrEnum):
 CONCENTRATION_MILLIGRAMS_PER_CUBIC_METER: Final = "mg/m³"
 CONCENTRATION_MICROGRAMS_PER_CUBIC_FOOT: Final = "μg/ft³"
 CONCENTRATION_PARTS_PER_CUBIC_METER: Final = "p/m³"
+CONCENTRATION_GRAMS_PER_LITER: Final = "g/L"
```
This adds new global units to Home Assistant. This requires an architectural discussion.

Alternatively, don't add this globally.

frenck commented on homeassistant/components/tuya/switch.py:
```diff
@@ -713,6 +713,27 @@
+            key=DPCode.BOOST_MODE,
+            translation_key="boost_mode",
+        ),
+        SwitchEntityDescription(
+            key=DPCode.DATA_TIP,
+            translation_key="data_tip",
+            entity_category=EntityCategory.CONFIG,
+        ),
```
Could you describe what this is?

frenck commented on homeassistant/components/tuya/switch.py:
```diff
@@ -713,6 +713,27 @@
+            translation_key="data_tip",
+            entity_category=EntityCategory.CONFIG,
+        ),
+        SwitchEntityDescription(
+            key=DPCode.TIP_SHOW,
+            translation_key="tip_show",
+            entity_category=EntityCategory.CONFIG,
+        ),
```
Same, could you explain what this means?

frenck commented on homeassistant/components/tuya/sensor.py:
```diff
@@ -1168,6 +1214,48 @@ class TuyaSensorEntityDescription(SensorEntityDescription):
+        TuyaSensorEntityDescription(
+            key=DPCode.OUTPUT,
+            translation_key="output_power",
+            device_class=SensorDeviceClass.POWER_FACTOR,
```
This seems a mismatch, what is the unit used in this case?

frenck commented on homeassistant/components/tuya/sensor.py:
```diff
@@ -160,6 +163,49 @@ class TuyaSensorEntityDescription(SensorEntityDescription):
+        TuyaSensorEntityDescription(
+            key=DPCode.CF_CURRENT,
+            translation_key="calcium_factor",
+            native_unit_of_measurement=CONCENTRATION_PARTS_PER_MILLION,
```
Why this using a fixed unit and not the unit provided by the API?

frenck commented on homeassistant/components/tuya/sensor.py:
```diff
@@ -160,6 +163,49 @@ class TuyaSensorEntityDescription(SensorEntityDescription):
+            key=DPCode.ORP_CURRENT,
+            translation_key="oxygen_reducing_potential",
+            device_class=SensorDeviceClass.VOLTAGE,
+            native_unit_of_measurement=UnitOfElectricPotential.MILLIVOLT,
```
Same, why is this a fixed unit?

frenck reviewed (changes requested):
Some things I noticed in most cases:

- Translation keys are set, but could not find the translation strings.
- Lot of added descriptions have a fixed/forced unit of measurement. Why wouldn't we be using the unit provided by the API instead?
- "qt" devices have been left out scope by Tuya originally. These devices are considered unsupported / outdated.
